From b7f406379370042d79fd07c976640fb772256baf Mon Sep 17 00:00:00 2001
From: Rob Clark <rclark@redhat.com>
Date: Wed, 17 Dec 2014 21:38:44 -0500
Subject: [drm] radeon: Add ability to get and change dpm state when radeon PX card is turned off

Message-id: <1418852380-13061-34-git-send-email-rclark@redhat.com>
Patchwork-id: 102174
O-Subject: [RHEL7 drm 33/89] drm/radeon: Add ability to get and change dpm state when radeon PX card is turned off
Bugzilla: 1173317
RH-Acked-by: Jerome Glisse <jglisse@redhat.com>
RH-Acked-by: Dave Airlie <airlied@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1173317
Upstream: since v3.16.4

commit 76325a92203b7db278a88d4e676de118bd4345f2

Author:     Pali Rohar <pali.rohar@gmail.com>
AuthorDate: Mon Aug 11 19:01:58 2014 +0200
Commit:     Rob Clark <rclark@redhat.com>
CommitDate: Wed Dec 10 17:16:32 2014 -0500

    drm/radeon: Add ability to get and change dpm state when radeon PX card is turned off

    commit b07a657e3a05b81c8a30d60e3f3746ca5a48ee62 upstream.

    This fixing commit 4f2f203976964e267dc477de6648bdb3acd2b74b

    bug:
    https://bugzilla.kernel.org/show_bug.cgi?id=76321

    Signed-off-by: Pali Rohar <pali.rohar@gmail.com>
    Signed-off-by: Alex Deucher <alexander.deucher@amd.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>

Signed-off-by: Rob Clark <rclark@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/gpu/drm/radeon/radeon_pm.c b/drivers/gpu/drm/radeon/radeon_pm.c
index a18596e..cca18e5 100644
--- a/drivers/gpu/drm/radeon/radeon_pm.c
+++ b/drivers/gpu/drm/radeon/radeon_pm.c
@@ -460,10 +460,6 @@ static ssize_t radeon_get_dpm_state(struct device *dev,
  struct radeon_device *rdev = ddev->dev_private;
  enum radeon_pm_state_type pm = rdev->pm.dpm.user_state;
 
- if  ((rdev->flags & RADEON_IS_PX) &&
-      (ddev->switch_power_state != DRM_SWITCH_POWER_ON))
-  return snprintf(buf, PAGE_SIZE, "off\n");
-
  return snprintf(buf, PAGE_SIZE, "%s\n",
    (pm == POWER_STATE_TYPE_BATTERY) ? "battery" :
    (pm == POWER_STATE_TYPE_BALANCED) ? "balanced" : "performance");
@@ -477,11 +473,6 @@ static ssize_t radeon_set_dpm_state(struct device *dev,
  struct drm_device *ddev = dev_get_drvdata(dev);
  struct radeon_device *rdev = ddev->dev_private;
 
- /* Can't set dpm state when the card is off */
- if  ((rdev->flags & RADEON_IS_PX) &&
-      (ddev->switch_power_state != DRM_SWITCH_POWER_ON))
-  return -EINVAL;
-
  mutex_lock(&rdev->pm.mutex);
  if (strncmp("battery", buf, strlen("battery")) == 0)
   rdev->pm.dpm.user_state = POWER_STATE_TYPE_BATTERY;
@@ -495,7 +486,12 @@ static ssize_t radeon_set_dpm_state(struct device *dev,
   goto fail;
  }
  mutex_unlock(&rdev->pm.mutex);
- radeon_pm_compute_clocks(rdev);
+
+ /* Can't set dpm state when the card is off */
+ if (!(rdev->flags & RADEON_IS_PX) ||
+     (ddev->switch_power_state == DRM_SWITCH_POWER_ON))
+  radeon_pm_compute_clocks(rdev);
+
 fail:
  return count;
 }
-- 
1.7.1